File: haskell/day11_part1_2017.hs

import System.IO

abs' x = if x < 0 then -x else x

max' a b = if a > b then a else b

distance x y z = (abs' x + abs' y + abs' z) `div` 2

main = do
    contents <- readFile "input.txt"
    let directions = words (map (\c -> if c == ',' then ' ' else c) contents)
        move "n"  (x, y, z) = (x, y+1, z-1)
        move "ne" (x, y, z) = (x+1, y, z-1)
        move "se" (x, y, z) = (x+1, y-1, z)
        move "s"  (x, y, z) = (x, y-1, z+1)
        move "sw" (x, y, z) = (x-1, y, z+1)
        move "nw" (x, y, z) = (x-1, y+1, z)
        updatePos (x, y, z) dir = move dir (x, y, z)
        (x, y, z) = foldl (\pos dir -> updatePos pos dir) (0, 0, 0) directions
    print (distance x y z)

File: nim/day11_part1_2017.nim
import strutils

type Direction = enum
  n, ne, se, s, sw, nw

proc move(x, y: var int, dir: Direction) =
  case dir
  of n: y.inc
  of ne: x.inc; y.inc
  of se: x.inc; y.dec
  of s: y.dec
  of sw: x.dec; y.dec
  of nw: x.dec; y.inc

proc distance(x, y: int): int =
  let dx = abs(x)
  let dy = abs(y)
  dx + max(0, (dy - dx) div 2)

proc main() =
  let input = readFile("input.txt").strip(chars={' '}).split(',')
  var x, y = 0
  for dir in input:
    case dir
    of "n": move(x, y, n)
    of "ne": move(x, y, ne)
    of "se": move(x, y, se)
    of "s": move(x, y, s)
    of "sw": move(x, y, sw)
    of "nw": move(x, y, nw)
    else: raise newException(IOError, "Invalid direction")
  echo "Minimum steps: ", distance(x, y)

when isMainModule:
  main()